File: apps/web/stores/nearestRain.ts
import type { CompassDirection, NearestRainResponse } from "@local-rain/shared";
import { defineStore } from "pinia";

interface NearestRainState {
  distance: number;
  eta: number;
  direction: CompassDirection | null;
  confidence: number;
  explanation: string;
  advice: string;
  hasRain: boolean;
  rainLatitude: number | null;
  rainLongitude: number | null;
  motionDirection: CompassDirection | null;
  speedKmh: number;
  approaching: boolean;
  previousDistance: number | null;
  rainChance: string;
  rainChancePct: number;
  rainIn1h: boolean;
  rainIn2h: boolean;
  loading: boolean;
  error: string | null;
}

const initialState = (): NearestRainState => ({
  distance: -1,
  eta: 0,
  direction: null,
  confidence: 0,
  explanation: "",
  advice: "",
  hasRain: false,
  rainLatitude: null,
  rainLongitude: null,
  motionDirection: null,
  speedKmh: 0,
  approaching: false,
  previousDistance: null,
  rainChance: "none",
  rainChancePct: 0,
  rainIn1h: false,
  rainIn2h: false,
  loading: false,
  error: null,
});

export const useNearestRainStore = defineStore("nearestRain", {
  state: initialState,
  getters: {
    distanceKm(state): number | null {
      if (!state.hasRain || state.distance < 0) return null;
      return state.distance / 1000;
    },
    distanceLabel(state): string {
      if (!state.hasRain || state.distance < 0) return "—";
      if (state.distance >= 1000) return `${(state.distance / 1000).toFixed(1)} km`;
      return `${Math.round(state.distance)} m`;
    },
    etaLabel(state): string {
      if (!state.hasRain || !state.approaching || state.eta <= 0) return "—";
      return `${state.eta} min`;
    },
  },
  actions: {
    setLoading(loading: boolean) {
      this.loading = loading;
    },
    setError(error: string | null) {
      this.error = error;
      this.loading = false;
    },
    setResult(payload: NearestRainResponse) {
      this.distance = payload.distance;
      this.eta = payload.eta;
      this.direction = payload.direction;
      this.confidence = payload.confidence;
      this.explanation = payload.explanation;
      this.advice = payload.advice;
      this.hasRain = payload.hasRain;
      this.rainLatitude = payload.rainLatitude;
      this.rainLongitude = payload.rainLongitude;
      this.motionDirection = payload.motionDirection;
      this.speedKmh = payload.speedKmh;
      this.approaching = payload.approaching;
      this.previousDistance = payload.previousDistance;
      this.rainChance = payload.rainChance ?? "none";
      this.rainChancePct = payload.rainChancePct ?? 0;
      this.rainIn1h = payload.rainIn1h ?? false;
      this.rainIn2h = payload.rainIn2h ?? false;
      this.loading = false;
      this.error = null;
    },
    reset() {
      Object.assign(this, initialState());
    },
  },
});

File: apps/api/app/schemas/nearest_rain.py
from pydantic import BaseModel, ConfigDict, Field

from app.utils.geo import CompassDirection


class NearestRainResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    distance: float
    eta: int = 0
    direction: CompassDirection
    confidence: int = Field(ge=0, le=100)
    explanation: str
    advice: str = ""
    has_rain: bool = Field(alias="hasRain")
    rain_latitude: float | None = Field(default=None, alias="rainLatitude")
    rain_longitude: float | None = Field(default=None, alias="rainLongitude")
    motion_direction: CompassDirection | None = Field(default=None, alias="motionDirection")
    speed_kmh: float = Field(default=0, alias="speedKmh")
    approaching: bool = False
    previous_distance: float | None = Field(default=None, alias="previousDistance")
    rain_chance: str = Field(default="none", alias="rainChance")
    rain_chance_pct: int = Field(default=0, alias="rainChancePct", ge=0, le=100)
    rain_in_1h: bool = Field(default=False, alias="rainIn1h")
    rain_in_2h: bool = Field(default=False, alias="rainIn2h")
    raining_here: bool = Field(default=False, alias="rainingHere")
    radar_timestamp: str | None = Field(default=None, alias="radarTimestamp")
    radar_age_minutes: int = Field(default=0, alias="radarAgeMinutes", ge=0)
    sky_state: str = Field(default="clear", alias="skyState")
    cloud_cover_pct: int = Field(default=0, alias="cloudCoverPct", ge=0, le=100)


class RainVectorItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    latitude: float
    longitude: float
    to_latitude: float = Field(alias="toLatitude")
    to_longitude: float = Field(alias="toLongitude")
    speed_kmh: float = Field(alias="speedKmh")
    direction: CompassDirection
    dbz: float = Field(ge=0, le=75)


class RainVectorsResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    vectors: list[RainVectorItem]
    generated_at: str = Field(alias="generatedAt")
